### ISSUE TITLE
[Bug] Saving a datasource does not update actions

### ISSUE THREAD

**Issue by Nikhil-Nandagopal** [appsmithorg/appsmith#2373]
## Description

On adding bindings to a datasource and saving it, the same is not reflected in the actions related to the datasource till the page is refreshed

### Steps to reproduce the behaviour:
1. Create an API and save a datasource against it
2. Navigate to the datasource and add a binding in the header.  ex. token: {{appsmith.store.token}}
3. save the datasource
4. Inspect the network
5. navigate to the API and run it.
6. The API runs without the binding
7. Refresh the page
8. Navigate to the API and run it.
9. The API runs with the binding

The API should run with the binding in both situations. This is happening because the action is updated on the server when the datasource is saved but not on the client

**Comment by sumitsum** [comment]
As per my investigation, action is not updated in the db with the header key value pairs when the API datasource is saved with header info. Removing the front-end tag.

**Comment by Nikhil-Nandagopal** [comment]
@sumitsum if that is the case why does refreshing the page work?

**Comment by sumitsum** [comment]
@Nikhil-Nandagopal it didn't work for me. Please check the video and see if I am looking at the right problem: 

https://user-images.githubusercontent.com/1757421/113002469-6d989880-918f-11eb-876c-0b113b4ae69b.mov

**Comment by Nikhil-Nandagopal** [comment]
@sumitsum there is no binding in your header as mentioned in the steps

**Comment by sumitsum** [comment]
@Nikhil-Nandagopal is the API header only supposed to be updated if the value comes from a binding and not other wise ?

**Comment by sumitsum** [comment]
@Nikhil-Nandagopal I have tried to used a binding in the header here. I am not able to see the headers getting updated. Please check if I this steps look ok: 

https://user-images.githubusercontent.com/1757421/113009823-eef32980-9195-11eb-93f5-12fafe0c2d36.mov

**Comment by hetunandu** [comment]
@Nikhil-Nandagopal, I don't think we should support this feature any more in this state. From my understanding, the use case is for custom login, and we should find a better way to enable this instead of building more on top of this unofficial feature. My suggestion is that we either create a new UX flow to allow for custom logins or if we want to enable bindings in datasources, we do it with the limited scope as today bindings are only at a page level while datasources are at an org level. AFAIK, the later is only possible in rest API today and it is not documented correctly as well.

**Comment by Nikhil-Nandagopal** [comment]
@hetunandu I believe this already works. The bug is that it only works after you refresh the page. So seems like a small fix isn't it?

**Comment by mohanarpit** [comment]
Can I ask what's the use-case for supporting bindings in the datasource? A datasource is shared with all apps in an organization. Having bindings in the datasource can lead to subtle bugs/issues where users end up binding values that aren't available in other apps. That's why I'm asking what the use-case is here.

**Comment by Nikhil-Nandagopal** [comment]
@mohanarpit it's for sending the access token as Hetu mentioned.

**Comment by mohanarpit** [comment]
@hetunandu Is it possible for us to limit what we bind based on different scenarios? For example, in a datasource, a user can only bind `{{env.key}}` or `{{appsmith.store.key}}`?

**Comment by hetunandu** [comment]
Yes, we can limit the scope for datasources. We don't have any scope at the org level. `appsmith.store` is app-level. So including this datasource in another app will still confuse.

**Comment by somangshu** [comment]
Another user probably faced the same issue, Here are the details

1. setup a rest datasource which returns a non-200 status code, like 401 or 404
2. put some key/values in the headers
RUN
3. remove headers
RUN
look in the error message and you see that the headers are still transmitted
it is simply not possible to remove them

The headers from the previous call are still present here until the page is refreshed. 
![unknown (3)](https://user-images.githubusercontent.com/11089579/129374511-dfa6b2fb-ed21-4768-af45-5b5566dbfe45.png)

@hetunandu @mohanarpit Is this a related issue?
